node/lib/util/merge_common.js

```diff
@@ -0,0 +1,317 @@
+                editMessage) {
+        assert.instanceOf(metaRepo, NodeGit.Repository);
+        if (null !== ourCommit) {
+            assert.instanceOf(ourCommit, NodeGit.Commit);
+        }
+        assert.instanceOf(theirCommit, NodeGit.Commit);
+        assert.isNumber(mode);
+        assert.isNumber(wdType);
+        if (null !== commitMessage) {
+            assert.isString(commitMessage);
+        }
+        assert.isFunction(editMessage);
+        this.d_metaRepo = metaRepo;
+        this.d_ourCommit = ourCommit;
+        this.d_theirCommit = theirCommit;
+        this.d_mode = mode;
+        this.d_wdType = wdType;
+        this.d_commitMessage = commitMessage;
+        this.d_editMessage = editMessage;
+        this.d_opener = new Open.Opener(metaRepo, null);
+        this.d_changeIndex = null;
+        this.d_changes = null;
+        this.d_conflictsMessage = "";
+    }
+
+    /**
+     * @property {Boolean} forceBare if working directory is disabled
+     */
+    get forceBare() {
+        return Open.WDTYPE.FORCE_BARE === this.d_wdType;
+    }
+
+    /**
+     * @property {NodeGit.Repository}
+     */
+    get metaRepo() {
+        return this.d_metaRepo;
+    }
+
+    /**
+     * @property {Opener}
+     */
+    get opener() {
+        return this.d_opener;
+    }
+
+    /**
+     * @property {NodeGit.Commit}
+     */
+    get theirCommit() {
+        return this.d_theirCommit;
+    }
+
+    /**
+     * @property {Open.WDTYPE}
+     */
+    get wdType() {
+        return this.d_wdType;
+    }
+
+    /**
+     * @property {MODE}
+     */
+    get mode() {
+        return this.d_mode;
+    }
+}
+
+/**
+ * @async
+ * @return {Object} return from sub name to `SubmoduleChange`
+ * @return {Object} return.simpleChanges from sub name to `Submodule`
+ * @return {Object} return.changes from sub name to `Submodule`
+ * @return {Object} return.conflicts from sub name to `Conflict`
+ */
+MergeContext.prototype.getChanges = co.wrap(function *(){
+    if (null === this.d_changes) {
+        this.d_changes =
+            yield CherryPickUtil.computeChangesBetweenTwoCommits(
+                this.d_metaRepo,
+                yield this.getChangeIndex(),
+                yield this.getOurCommit(),
+                this.d_theirCommit);
+    }
+    return this.d_changes;
+});
+
+/** 
+ * @async
+ * @return {NodeGit.Commit} return left side merge commit
+ */
+MergeContext.prototype.getOurCommit = co.wrap(function *() {
+    if (null !== this.d_ourCommit) {
+        return this.d_ourCommit;
+    }
+    if (this.forceBare) {
+        throw new UserError("Left side merge commit is undefined!");
+    }
+    this.d_ourCommit = yield this.d_metaRepo.getHeadCommit();
+    return this.d_ourCommit;
+});
+
+/**
+ * return an index object that contains the merge changes and whose tree
+ * representation will be flushed to disk.
+ * @async
+ * @return {NodeGit.Index} 
+ */
+MergeContext.prototype.getIndexToWrite = co.wrap(function *(){
```

Comment:
In the bare mode, the `index` comes from  [`Merge.commits`](https://libgit2.org/libgit2/#HEAD/group/merge/git_merge_commits), because we cannot assume there is an index object in this mode. 

In the interactive mode, the `index` object is read from the current repo, and then we can add edited files later with [`index.addByPath(...)`](https://github.com/twosigma/git-meta/blob/master/node/lib/util/merge_util.js#L230)